### DISCUSSION TITLE
bulk insert of objects and related objects

### DISCUSSION THREAD

**Discussion by Et7f3** [sqlalchemy/sqlalchemy#11002] [Usage Questions] [answered]
Hello,

Here is the minimal reproduction of my issue.

My models:

```python
class Post(Base):
  __tablename__ = "posts"

  id: Mapped[int] = mapped_column(primary_key=True, autoincrement="auto", init=False)
  text: Mapped[str]
  comments: Mapped[List["Comment"]] = relationship(backref="post")

class Comment(Base):
  __tablename__ = "comment"

  id: Mapped[int] = mapped_column(primary_key=True, autoincrement="auto", init=False)
  post_id: Mapped[int] = mapped_column(ForeignKey("posts.id"), init=False)
  text: Mapped[str]
```

I have tried with `backred` and `back_populates`

My insert code:

```python
with open_session() as db:
  # I expect at least one syntax works
  db.execute(insert(Post), [{ "text": "My first post", "comments": [ { "text": "My first comment" } ] } ])
  db.execute(insert(Post), [{ "text": "My second post", "comments": [ Comment(text="My second comment") ] }])
  cmt = Comment(text="My third comment")
  cmt.post_id = 7 # An available id in posts database
  db.execute(insert(Post), [{ "text": "My third post", "comments": [ cmt ] }])
```

Expected result:
When I use one of the syntax above (or another I didn't found in the docs): The post is added and the comment also.

Actual result:
Only Post is aded without commend. I thought it was because it couldn't guess id so in the last try I hard coded a valid id and even with this it wasn't in my db.

A workaround I could think of is to use bulk only for post without comments and for post with comments I do a regular insert get the returned id and is use to create a list of comments I could add after via bulk insert.

I don't mind slow query or multiple db hit (I was just lazy to think I could just push my model to db and do all the dirty works for me)

The closest I have found: https://stackoverflow.com/a/59121078/7227940

If we dig deeper we can have a [more clever solution](https://www.depesz.com/2008/03/20/getting-multiple-values-from-sequences/#comment-25483) (but can't seems to find python equivalent)

**Comment by CaselIT** [comment, marked as answer]
Hi,

That's not supported at the moment, and I don't think it would be a clear abstraction, since it would be very strange if `execute(insert(some_class), ...)` were to execute inserts on other unrelated tables.

To use relationships you can use use objects and add them to a session https://docs.sqlalchemy.org/en/20/tutorial/orm_related_objects.html#persisting-and-loading-relationships

**Reply by Et7f3** [reply]
Hi,

Thanks for your quick reply.

> were to execute inserts on other unrelated tables.

I agree but strange behaviour can always be guarded with an optional kwarg like insert_dependants=True. So I hoped one exists and I haven't saw it but you confirm it doesn't.

> To use relationships you can use use objects and add them to a session

Ok so it is like what I found on SO.


Do you know how it computes id (6) even if object is still in transient state: https://docs.sqlalchemy.org/en/20/tutorial/orm_related_objects.html#cascading-objects-into-the-session

**Reply by CaselIT** [reply]
Well if the ids are generated by the db the only way of getting them is to flush the data to the db.
If the id generation is client side it depends on how it's generated:
- if you are using dataclasses using default_factory, that's generated by the init
- if you use a default outside dataclasses or [insert_default](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column.params.insert_default) that's generated during a flush

in your cases you are using autoincrement, so the ids are a computed by the db and will be available after a flush

**Reply by Et7f3** [reply]
In the example an user object is created and it's attribute is updated to add an address. I have played a bit and it seems that you can directly construct object with related object. You can then add object to the session.

This works:
```python
  db.add_all([
    Post("My first post", [ Comment("My first comment") ]),
    Post("My second post", [ Comment("My second comment"), Comment("My second second comment") ])
  ])
```

The object is created in database with the associated comments.

PRO:
- works as intended (you can dump a model to db and it works)
CON:
- We need to create explicitly the objects: even though we don't need to read back from db

When we compare with my first attempt `db.execute(insert(Post), [{ "text": "My first post", "comments": [ { "text": "My first comment" } ] } ])`

We say that root element is Post and by following the model we can see "comments" is a Comment. So a quick fix would be to convert this list of bulk insert to object with related children object.
cleaner fix (because use less memory) build the query the same way that session_add do the jobs.

I have implemented the quickfix method (specialized to my models) in my code and it works.

**Comment by HMaker** [comment]
`Session.add_all()` is at the ORM layer and abstracts away many things, it handles insertions in multiple tables by correctly mapping object dependencies (Post's comments in this case)

`db.execute(insert(Post), [{ "text": "My first post", "comments": [ { "text": "My first comment" } ] } ])` is a lower level construct, it's a direct insert in an specific table, `Post`. I think here SQLAlchemy handles only the target table and all possible parents in case of joined table inheritance. It won't insert the comments.

You can do 2 calls to `db.execute()`, the first one would insert all posts and retrieve its primary keys
```py
posts = session.scalars(insert(Post).returning(Post), [{ "text": "My first post" } ])
session.execute(insert(Comment), [{ "text": "My first comment", "post_id": posts[0].id } ])
```

You will aways need 2 insert statements (one for the Post and another for the Comments table) even if the post's primary key is generated client-side.
